Gruf 2.0

splittingred opened the pull request:
## What? Why?

Gruf 2.0 is a *major* version bump, and has many backwards-incompatible changes. 

This version:
* Moves away from writing and declaring method implementations in the generated server stub, and now allows you to "bind" a Controller to a service to give you a pristine class to work on
* Restricts various APIs that are no longer needed
* Adds a cleaner API for developing RPC methods in a `Gruf::Controllers::Base` class and sets up framework for future feature development
* Adds more inject-ability into the `Gruf::Server` class
* Removes hooks. They are now replaced by a pure `Gruf::Interceptors::ServerInterceptor` class, which paves the way for:
  * Deprecating the arbitrary difference between authentication, instrumentation, and generic hooks
  * Delegated complete control of the order of execution of various interceptors
  * A more controlled API for interceptors that has a view of a controller request "context"

## New Controllers

Controllers now bind to services:

```ruby
class ThingController < ::Gruf::Controllers::Base
  bind ::Rpc::ThingService::Service

  def get_thing
    thing =  Rpc::Thing.new(id: request.message.id, name: 'Foo')
    Rpc::GetThingResponse.new(thing: thing)
  end
end
```

You'll also notice a new `request` object that contains the `message` object, as well as other things (the service class, active call, metadata, etc).

These controllers are also fully thread-safe, as opposed to the old `Gruf::Service`, which was shared between calls.

## How was it tested?

* `rspec`
* `spec/demo_server` then executing various rake commands
* Using in [gruf-demo](https://github.com/bigcommerce/gruf-demo) - see [releases/2.x branch](https://github.com/bigcommerce/gruf-demo/tree/releases/2.x)

junedkazi commented on README.md:
```diff
@@ -8,17 +8,18 @@ provide a more streamlined integration into Ruby and Ruby on Rails applications.
-gruf currently has active support for gRPC 1.4.x. gruf is compatible and tested with with Ruby 2.2, 2.3, and 2.4. gruf 
-is also not [Rails](https://github.com/rails/rails)-specific, and can be used in any Ruby framework (such as 
-[Grape](https://github.com/ruby-grape/grape), for instance).
+gruf currently has active support for gRPC 1.4.x-1.6x. gruf is compatible and tested with with Ruby 2.2, 
```
2 `with`

junedkazi replied:
should it be `1.6.x` ?

billthompson commented on README.md:
````diff
@@ -308,7 +269,7 @@ It supports formatters (including custom ones) that you can use to specify the f
 ```ruby
 Gruf.configure do |c|
-  c.instrumentation_options[:request_logging] = {
+  c.interceptors[Gruf::Interceptors::Instrumentation::RequestLogging::Interceptor] = {
````
This style of registering interceptors feels like it gives a little too much power to the consumers. From this (I haven't looked at the implementation yet), I think is probably using a hash to guarantee uniqueness. The problem with this approach is when interceptor order becomes important.

I think we can get more flexibility to change and new features by following a process like Rack::Middleware does:
```
c.interceptors.use Gruf::Interceptors::Instrumentation::RequestLogging::Interceptor, options: { formatter: :logstash}
```

Then, in the future, we can add the ability to control order without changing the underlying data structure via something like:
```
c.interceptors.insert_before ...
c.interceptors.insert_after ...
```

What do you think?

splittingred replied:
I thought about this. So in Ruby insertion order is maintained in Hash ([see here](https://www.igvita.com/2009/02/04/ruby-19-internals-ordered-hash/)).

I understand the idea here though, and agree with you that implicit behavior can be a bit confusing. I'll look into this.

billthompson replied:
Awesome! Thanks for providing a link on hash insertion order and looking into changing the API.

billthompson commented on spec/pb/thing_controller.rb:
```diff
@@ -15,65 +15,68 @@
 
-class ThingService < ::Rpc::ThingService::Service
-  include Gruf::Service
+class ThingController < ::Gruf::Controllers::Base
```
Can we provide a more concrete example in the specs and documents? I feel like `Thing*` is just too abstract to easily comprehend. I'd like to see us use `User*`, as most developers can relate to that or have built systems with users. While the code will be the same, I think it will be easier to apply to the developer's situation using a less abstract concept.

splittingred replied:
Pushing to gruf 2.1

billthompson commented on UPGRADING.md:
```diff
@@ -1,5 +1,117 @@
+
+Interceptors behave very similarly to outer around hooks, with the following changes:
+
+* They are now provided an instance of `Gruf::Controllers::Request`, which contains request information
```
🌟

billthompson commented on bin/gruf:
```diff
@@ -17,13 +17,20 @@
 rescue => e
-  Gruf.logger.fatal "FATAL ERROR: #{e.message} #{e.backtrace.join("\n")}"
+  msg = "FATAL ERROR: #{e.message} #{e.backtrace.join("\n")}"
+  if Gruf.logger
```
Can Gruf.logger be a `Logger.new(STDOUT)` if one is not provided during configuration? That would allow us to omit this conditional here and simplify logging in general.

splittingred replied:
Yup, it is. `self.logger = ::Logger.new(STDOUT)` in configuration.rb.

However, this conditional here catches it on the chance it fatals before that logger is setup.

billthompson commented on lib/gruf/configuration.rb:
```diff
@@ -85,27 +82,24 @@ def reset
       VALID_CONFIG_KEYS.each do |k, v|
         send((k.to_s + '='), v)
       end
+      self.root_path = Rails.root.to_s.chomp('/') if defined?(Rails)
```
🍺 It would probably better to have a `gruf-rails` gem to handle framework specific things like this in the future.

billthompson commented on lib/gruf/controllers/base.rb:
```diff
@@ -0,0 +1,103 @@
+          method_key = name.to_s.underscore.to_sym
+
+          service.class_eval do
+            if rpc_desc.request_response?
```
🍺 It would be nice to encapsulate this logic into a `Gruf::Controllers::MethodFactory` (or something like that)

billthompson commented on lib/gruf/controllers/request.rb:
```diff
@@ -15,51 +15,75 @@
+      # @return Enumerable<Object>
+      #
+      def messages
+        if type == :client_streamer
```
🍺 IMHO, this class is asking for a `RequestType` object to manage the logic of working with a specific request type. That could help this class offload the task conditional around L40 in `type` by building the `RequestType` object from the `rpc_desc` and also delegate the formatting of the messages to the RequestType class instead of doing that here.

I understand this may be considered as either an SRP violation or a premature abstraction so that's why I'm making this suggestion as optional.

If you need any clarification on this suggestion just let me know.

billthompson commented on lib/gruf/errors/handlers.rb:
```diff
@@ -0,0 +1,77 @@
+      # @param [Hash] metadata (Optional) Any metadata to inject into the trailing metadata for the response
+      #
+      def fail!(error_code, app_code = nil, message = '', metadata = {})
+        e = error
```
🍺 Build the error using a factory/builder pattern since there are transformations and conditional logic for the error state not specific to the handler logic.

splittingred replied:
Pushing to 2.1, as this doesn't affect any public contracts, and I'm not sure we want to do this as it's just a field mapping

billthompson commented on lib/gruf/errors/handlers.rb:
```diff
@@ -0,0 +1,77 @@
+#
+module Gruf
+  module Errors
+    module Handlers
```
⭐️ for making `fail` and `add_field_error` public 🙇

billthompson commented on lib/gruf/errors/handlers.rb:
```diff
@@ -0,0 +1,77 @@
+      # @param [String] message A given error message for the field
+      #
+      def add_field_error(field_name, error_code, message = '')
+        error.add_field_error(field_name, error_code, message)
```
This makes me feel like `@error` should be an instance variable on the controller instance and this module just exposes convenience methods for interacting with it. Not a blocker, but would be nice to have I think as the controller should be responsible for managing the error state. And then this is more of an `Errors::Helpers` mixin instead of a handler.

splittingred replied:
To do this, I'm going to actually pass that instance variable into the interceptor initializers as well, so interceptors will be able to modify error state. I think this is useful for a few reasons - allowing interceptors to "fail gracefully" and handle multiple interceptors failing but not exiting early; also to allow field-level validation to move upstream without necessarily breaking out of the request cycle prior to the main method call.

billthompson replied:
That's extremely powerful! Good call.

billthompson commented on lib/gruf/interceptors/context.rb:
```diff
@@ -15,30 +15,41 @@
-
-        rc.active_call.output_metadata.update(metadata_key => rc.execution_time.to_s)
+      def initialize(request, interceptors = [])
+        @interceptors = interceptors.map { |klass, opts| klass.new(request, opts) }
```
🍺 `Array(interceptors).map...`

billthompson commented on lib/gruf/interceptors/instrumentation/output_metadata_timer.rb:
```diff
@@ -15,29 +15,34 @@
-            route_key = payload.fetch(:method, 'unknown')
-            body = payload.fetch(:message, '')
+        def call
+          return unless request.active_call && request.active_call.respond_to?(:output_metadata)
```
Since `nil` responds to `respond_to?`, you can get away with just `request.active_call.respond_to?(:output_metadata)` here.

billthompson commented on lib/gruf/interceptors/instrumentation/output_metadata_timer.rb:
```diff
@@ -15,29 +15,34 @@
+          result = Gruf::Interceptors::Timer.time do
+            yield
           end
+          request.active_call.output_metadata.update(metadata_key => result.elapsed.to_s)
```
💁‍♂️ Law of Demeter here.

billthompson commented on lib/gruf/interceptors/instrumentation/request_logging/interceptor.rb:
```diff
@@ -0,0 +1,198 @@
+          #
+          # @return [String]
+          #
+          def call
```
🍺 This method is has a high cyclomatic complexity. Consider breaking this up.

splittingred replied:
Pushing to gruf 2.1

billthompson commented on lib/gruf/interceptors/instrumentation/request_logging/interceptor.rb:
```diff
@@ -0,0 +1,198 @@
+            end
+
+            payload = {}
+            if !client_streamer?
```
If you perform the suggested refactor to build a `RequestType` object, you can update this part of the code to as the `RequestType` directly if it is a `client_streamer?` instead of using the adhoc logic on L89.

billthompson commented on lib/gruf/server.rb:
```diff
@@ -64,44 +64,54 @@ def start!
+      path = File.realpath(controllers_path)
+      $LOAD_PATH.unshift(path)
+      Dir["#{path}/**/*.rb"].each do |f|
+        next if f.include?('_pb') # exclude if people include proto generated files in app/rpc
```
That is a pretty opinionated assumption based on naming convention - is it necessary? I'd prefer instead if we would inspect the file contents and only require it is it subclasses `::Gruf::Controllers::Base`. That would also protect us if someone includes controller concerns in the controller's directory - (see https://github.com/bigcommerce/gruf/pull/25)

splittingred replied:
Yes - due to the way gRPC generates and includes files, we need to ensure this file is not included (otherwise it will error out with a "Service already defined" error). Unfortunately this is an issue due to the way gRPC generates the Ruby files, and not something in Gruf.

billthompson replied:
Can we verify based on the contents of the file instead? That will allow us to ensure we only require Gruf::Controllers. The protobuf files won't be subclassing `::Gruf::Controllers::Base` so this method is the safest IMHO. The current loader will still load non-service files (such as ActiveSupport::Concerns) as mentioned in the previous PR.

splittingred replied:
1) This doesn't _run_ or mount anything in those, which the prior did; it just loads the files.
2) We _can't_ load the file due to the require calls in the generated files, and I don't particularly want to have to do file opening and string parsing on the file itself.

billthompson replied:
Ok. In our existing implementation I don't think we are using concerns anymore, but if this causes an issue I'll raise a new issue at that time.

splittingred commented:
@billthompson Thanks for the feedback! Addressed your comments: Cleaned up the error interfaces functionality, added a `Gruf::Controllers::Request::Type` class, and made a new `Gruf::Controllers::ServiceBinder` class for binding a controller to a service.

billthompson commented on lib/gruf/interceptors/base.rb:
```diff
@@ -0,0 +1,51 @@
+      # @var [Gruf::Controllers::Request] request
+      attr_reader :request
+      # @var [Gruf::Error] errors
+      attr_reader :errors
```
🍺 With a plural name, I would expect this to be a collection. What do you think about calling this just `error`?

splittingred replied:
Yeah I had moved it to `errors` as it's really a payload of potential errors, but thinking about it more it _is_ a "holistic" error that to the end client is a single "error". Renaming.

billthompson commented on lib/gruf/interceptors/registry.rb:
```diff
@@ -0,0 +1,137 @@
+      def interceptors_mutex(&block)
+        @interceptors_mutex ||= begin
+          require 'monitor'
+          Monitor.new
```
TIL 🙇

billthompson commented on lib/gruf/interceptors/registry.rb:
```diff
@@ -0,0 +1,137 @@
+      def insert_after(after_class, interceptor_class, options = {})
+        interceptors_mutex do
+          pos = nil
+          @registry.each_with_index do |o, idx|
```
🍺 This code can be simplified in both of the insert methods:
```
pos = @registry.find_index { |interceptor| interceptor.fetch(:klass, '') == after_class.to_s }
```

splittingred commented:
@billthompson 🙇 ♻️
